### PR TITLE
Replace mx by scaled

### LINKED ISSUE #1134: update MinHash constructor API in Rust to use `scaled` instead of `max_hash`
In #1128, we are deprecating the old `max_hash` parameter in favor of the new `scaled` parameter. They can be interconverted easily with no additional information (see functions `_get_max_hash_for_scaled` and `_get_scaled_for_max_hash`).

However, it is hard to fully remove `max_hash` from the Python API because it is used in the Rust API as well!
 See the call to `lib.kmerminhash_new` in `sourmash/minhash.py:MinHash.__init__` (currently in #1128 only).

So the goal of this issue is to swap out the `max_hash` parameter with the `scaled` parameter in the Rust API.

There are two places to do this in Rust - one is in the `kmerminhash_new` function, and the other is in the `KmerMinHash::new` code. I would suggest starting with the former and doing the conversion there, so that
we can get the API bridge between Python and Rust working first, but YMMV :)

Some suggestions for getting started -

* get a full Python and Rust development environment up and running per [the docs](https://sourmash.readthedocs.io/en/latest/developer.html)
* dig into the implementation of `kmerminhash_new`in `src/core/src/ffi/minhash.rs`
* add the scaled_to_max_hash and max_hash_to_scaled conversion functions into Rust
* swizzle the `mx` parameter over to `scaled` in `kmerminhash_new` and then convert it to `max_hash` before passing it into the `::new` method.

### PR DESCRIPTION
**Pull request by xmnlab**
- [ ] Is it mergeable?
- [ ] `make test` Did it pass the tests?
- [ ] `make coverage` Is the new code covered?
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

Fixes #1134

### CODE REVIEW

**Comment by xmnlab** [comment]
@ctb @luizirber it is still raising some errors. I will work on that again today.
is this change what you had in mind? Am I going in the right direction? thanks!

**Comment by ctb** [comment]
on a quick skim, this looks like the right direction to me!

**Comment by xmnlab** [comment]
thanks @ctb! I will try to have it ready for review today!

**Comment by ctb** [comment]
On Tue, Aug 04, 2020 at 07:56:43AM -0700, Ivan Ogasawara wrote:
> thanks @ctb! I will try to have it ready for review today!

no hurry, take your time :)

**Comment by ctb** [comment]
hi @xmnlab note that we just changed the default main branch in this repository to `latest` - at some point before merge, please update your base to match (click Edit, change base to `latest`). Thanks!

**Comment by luizirber** [comment]
Nice PR, @xmnlab!

But I think we need to be more explicit in the consequences of changing `max_hash` to `scaled` in the functions. My main concern is that current code using `max_hash` will continue "working" with `scaled` (since they are the same type), but the **meaning** of the parameter changed completely!

We can use this opportunity to bring it more in line with the changes happening on the Python side in #1128.
To make it break explicitly and force other crates using sourmash to do the changes, maybe define `new` as
```rust
pub fn new(
  scaled: u64,
  ksize: u32,
  hash_function: HashFunctions,
  track_abundance: bool,
  seed: u64,
  num: Option<u32>,
) -> KmerMinHash
```
and similarly in the FFI
```rust
#[no_mangle]
pub unsafe extern "C" fn kmerminhash_new(
    scaled: u64,
    k: u32,
    prot: bool,
    dayhoff: bool,
    hp: bool,
    track_abundance: bool,
    seed: u64,
    num: u32
) -> *mut SourmashKmerMinHash {
```

(let @ctb chime in before implementing these changes, @xmnlab =])

**Comment by xmnlab** [comment]
thanks @luizirber! I updated my branch on top of the "latest" branch. about your suggestions, sounds good for me, I am waiting for a green light from @ctb  :)

**Comment by ctb** [comment]
On Wed, Aug 05, 2020 at 05:42:04PM -0700, Ivan Ogasawara wrote:
> thanks @luizirber! I updated my branch on top of the "latest" branch. about your suggestions, sounds good for me, I am waiting for a green light from @ctb  :)

oh, don't listen to me, listen to @luizirber :).

**Comment by xmnlab** [comment]
@ctb I will work on that! thanks! :)

**Comment by xmnlab** [comment]
not sure if I am doing anything wrong, or if it is relevant, but rust `scaled_for_max_hash` and python `_get_scaled_for_max_hash` return a different value for number `1`

```
python result -> 18446744073709551616
rust result   -> 18446744073709551615 
```

if I change the python operation to use `//` instead of `/` it seems it returns the same rust result:

```
int(round(get_minhash_max_hash() // 1, 0))
18446744073709551615 
```

**Comment by luizirber** [comment]
> not sure if I am doing anything wrong, or if it is relevant, but rust `scaled_for_max_hash` and python `_get_scaled_for_max_hash` return a different value for number `1`
> 
> ```
> python result -> 18446744073709551616
> rust result   -> 18446744073709551615 
> ```

Good catch! That is definitely a bug, but we never triggered it because setting `max_hash=1` is not a viable use case. 

> if I change the python operation to use `//` instead of `/` it seems it returns the same rust result:
> 
> ```
> int(round(get_minhash_max_hash() // 1, 0))
> 18446744073709551615 
> ```

Instead of fixing the Python code, I think we should fix the Rust code (to guarantee it has the same behavior as any existing code using sourmash).

**Comment by xmnlab** [comment]
> Instead of fixing the Python code, I think we should fix the Rust code (to guarantee it has the same behavior as any existing code using sourmash).

this is the current rust implementation:

```rust
pub fn scaled_for_max_hash(max_hash: u64) -> u64 {
    match max_hash {
        0 => 0,
        _ => u64::max_value() / max_hash,
    }
}
```

maybe I am wrong, but as far as I understood, the rust result is the u64 max value and the python result is equal to the rust max value +1, so the rust call will break (integer overflow). does It make sense?

**Comment by luizirber** [comment]
> maybe I am wrong, but as far as I understood, the rust result is the u64 max value and the python result is equal to the rust max value +1, so the rust call will break (integer overflow). does It make sense?

You're correct, it was a trick question! (not really, I didn't notice that, and you analyzed correctly =])

In this case, let's keep `scaled_for_max_hash` in Rust as is.

**Comment by xmnlab** [comment]
Sounds pretty good! Do you want I change it here in this PR? Or should I
fix that in a new one?

**Comment by luizirber** [comment]
> Sounds pretty good! Do you want I change it here in this PR? Or should I fix that in a new one?

Here is good

**Comment by luizirber** [comment]
so, I just hit this in another PR, and.. I think the Rust function is broken =]

(it wasn't used in any place, and there are no tests, so...)

I tried checking the scaled for `184467440737095520`, and Python gives 100 while Rust says 99...

**Comment by xmnlab** [comment]
do you have any idea how to fix that? is there any scientific paper that describes this function?

**Comment by luizirber** [comment]
> do you have any idea how to fix that? 

Before thinking on how to fix it, this is the perfect example for using [property-based testing][2] (because we have two functions that applied in sequence should return the initial value). In Python we have [hypothesis][0] tests, and in Rust there is [proptest][1].

[0]: https://hypothesis.works/
[1]: https://github.com/AltSysrq/proptest
[2]: https://fsharpforfunandprofit.com/posts/property-based-testing/

In `tests/test__minhash_hypothesis.py`:
```python
 @given(st.integers(min_value=1, max_value=2**64 - 1))                                                                                                                                        
 def test_scaled_max_hash_conversion(scaled):                                                                                                                                                 
     max_hash = _get_max_hash_for_scaled(scaled)                                                                                                                                              
     new_scaled = _get_scaled_for_max_hash(max_hash)                                                                                                                                          
                                                                                                                                                                                              
     assert scaled == new_scaled  
```

in `src/core/tests/minhash.rs`:
```rust
proptest! {
#[test]
fn scaled_and_max_hash(scaled in u64::ANY) {
    let max_hash = max_hash_for_scaled(scaled).unwrap();
    let new_scaled = scaled_for_max_hash(max_hash);

    assert_eq!(scaled, new_scaled);
}
}
```

> is there any scientific paper that describes this function?

Ouch. Not yet =]

The main issue is that the definition of the function doesn't worry enough about numerical precision...
`max_hash` and `scaled` are the inverse of each other
`max_hash = max_value / scaled`
`scaled = max_value / max_hash`
where `max_value` is `(2 ** 64) - 1` for 64-bit integers.

But, that division is not well defined (but rounding floats is definitely not the solution! Probably proper [floor and ceil][3] operations). I'll think a bit about it and comment again later.

[3]: https://en.wikipedia.org/wiki/Floor_and_ceiling_functions

**Comment by luizirber** [comment]
OMG what a gigantic rabbit hole.

> But, that division is not well defined (but rounding floats is definitely not the solution! Probably proper [floor and ceil](https://en.wikipedia.org/wiki/Floor_and_ceiling_functions) operations). I'll think a bit about it and comment again later.

The main design constrains:
- We need to remain compatible with already existing scaled MinHash sketches
- Realistic scaled values are in the range (1, 100000)

From that:
- The tests I suggested point very quickly that it is hard to invert max_hash/scaled for scaled values outside the range (like `42537073439`)
- Fixing for values outside the range might mess values inside the range...
- Hypothesis tests in Python with the realistic range don't fail with the current Python impl.

So I guess we should be bug-compatible in Rust too.

Thoughs, @ctb?

**Comment by ctb** [comment]
My hot takes --

* bug compatibility is good, for now!
* is there a way to ensure that the round trip continues to work, as in `scaled == get_scaled_for_max_hash(get_max_hash_for_scaled(scaled))`, and each function returns integer values?
* some tests (especially ones that discover absurdities) would be most welcome

**Comment by xmnlab** [comment]
thanks @ctb and @luizirber for all the explanations and suggestions.

but I am not too sure how to move forward now ...

**Comment by luizirber** [comment]
>     * is there a way to ensure that the round trip continues to work, as in `scaled == get_scaled_for_max_hash(get_max_hash_for_scaled(scaled))`, and each function returns integer values?
> 
>     * some tests (especially ones that discover absurdities) would be most welcome

That's what the property tests are doing (it's implemented as you described in the first item, and hypothesis/proptest generate input data like the second one).

In the Hypothesis case we can be more explicit and provide some examples for the values we tend to use:
```python
@given(st.integers(min_value=1, max_value=100000))
@example(1)
@example(10)
@example(100)
@example(1000)
@example(2000)
@example(10000)
def test_scaled_max_hash_conversion(scaled):
    max_hash = _get_max_hash_for_scaled(scaled)
    new_scaled = _get_scaled_for_max_hash(max_hash)
    assert scaled == new_scaled
```

> thanks @ctb and @luizirber for all the explanations and suggestions.
> 
> but I am not too sure how to move forward now ...

How about:
- Add the hypothesis and proptest tests
- fix the Rust `scaled_for_max_hash` and `max_hash_to_scaled` to match the Python impl (using float numbers, and rounding)
- PROFIT! =P

**Comment by xmnlab** [comment]
great! I have a good picture now to start! thanks!

**Comment by xmnlab** [comment]
> fix the Rust scaled_for_max_hash and max_hash_to_scaled to match the Python impl (using float numbers, and rounding)

it seems it loses precision 

```rust
>>> println!("using f64: {}", u64::max_value() as f64);
>>> println!("using u64: {}", u64::max_value());

using f64: 18446744073709552000
using u64: 18446744073709551615
```

and casting to u64, it will truncate to the max of u64, so it will never be the same as python (for this example the result would be `18446744073709551616`):

```rust
>>> println!("using ceil: {}", ((u64::max_value() as f64) / 1.0f64).ceil() as u64);
>>> println!("using u64:  {}", u64::max_value());

using ceil: 18446744073709551615
using u64:  18446744073709551615
```

any idea about how we should handle this problem?

**Comment by luizirber** [comment]
> it seems it loses precision
> 
> ```rust
> >>> println!("using f64: {}", u64::max_value() as f64);
> >>> println!("using u64: {}", u64::max_value());
> 
> using f64: 18446744073709552000
> using u64: 18446744073709551615
> ```

yikes!

(is thix evcxr? Cool!)

> and casting to u64, it will truncate to the max of u64, so it will never be the same as python (for this example the result would be `18446744073709551616`):
>
> ```rust
> >>> println!("using ceil: {}", ((u64::max_value() as f64) / 1.0f64).ceil() as u64);
> >>> println!("using u64:  {}", u64::max_value());
> 
> using ceil: 18446744073709551615
> using u64:  18446744073709551615
> ```

the max value for 64 bits is 2**64 -1, which is `18446744073709551615`, so the first version (using ceil) is correct. I think you can use that for implementing the conversion?

**Comment by xmnlab** [comment]
> (is thix evcxr? Cool!)

yes! that is very helpful!

> I think you can use that for implementing the conversion?

but the value returned by the current python implementation is still different `18446744073709551616` ... to have the same result with the python implementation I would need to do `int(round(get_minhash_max_hash() // 1, 0))`

not sure if I am missing something

**Comment by luizirber** [comment]
> but the value returned by the current python implementation is still different `18446744073709551616` ... to have the same result with the python implementation I would need to do `int(round(get_minhash_max_hash() // 1, 0))`

`_get_max_hash_for_scaled(1)` is `18446744073709551615` here, I see `18446744073709551616` when doing `_get_scaled_for_max_hash(1)`.

Since it is out of the realistic scaled values range (I mean, a Scaled MinHash with `scaled=18446744073709551616` or `max_hash=1` will literally have only `(0, 1)` as possible values!), I don't think it is a big issue to leave as is (returning `18446744073709551615`).

(And even tho we said "bug-compatible", that is a bug we want to fix, because `18446744073709551616` is out of the range of possible 64-bit integers... It is easier to see this in hexadecimal:
```python
>>> "{:X}".format(18446744073709551616)
'10000000000000000'

>>> "{:X}".format(18446744073709551615)
'FFFFFFFFFFFFFFFF'
```
)

**Comment by xmnlab** [comment]
sorry .. I didn't have time too much last week. my vacations start on Monday so I will be back here tomorrow!! thanks for the patience :)

**Comment by luizirber** [comment]
some suggestion to move forward with this:

1) `_get_max_hash_for_scaled` and `_get_scaled_for_max_hash` are returning a value that doesn't fit in 64 bits in some cases. To fix that, we can do `return min(..., MINHASH_MAX_HASH)` to cap the value at the maximum value representable in 64-bit unsigned integers
2) There is another place in `sourmash/minhash.py` that is calling `lib.kmerminhash_new`, in `__setstate__`. If you use the same logic you already did in `__init__.py` it is going to work.

With these two changes the tests should be passing. And I'll add some other comments about the API, just a sec =]

File: src/core/src/ffi/minhash.rs
Comment (luizirber) [inline src/core/src/ffi/minhash.rs]: since we are changing the APIs, I think we can make `KmerMinHash::new` and `kmerminhash_new` look more like each other. Maybe
```rust
#[no_mangle]
pub unsafe extern "C" fn kmerminhash_new(
    scaled: u64,
    k: u32,
    hash_function: HashFunctions,
    seed: u64,
    track_abundance: bool,
    n: u32,
) -> *mut SourmashKmerMinHash {
```
and
```rust
impl KmerMinHash {
    pub fn new(
        scaled: u64,
        ksize: u32,
        hash_function: HashFunctions,
        seed: u64,
        track_abundance: bool,
        num: u32,
    ) -> KmerMinHash {
```

This will change how `lib.kmerminhash_new` is called in Python. more like:
```python
hash_function = lib.HASH_FUNCTIONS_MURMUR64_DAYHOFF # or DNA, HP, PROTEIN
self._objptr = lib.kmerminhash_new(
    scaled, ksize, hash_function, seed, track_abundance, n
)
```

Eventually we want to expose `HashFunctions` properly to Python (see #751 for discussion), but for now we can use `lib.HASH_FUNCTIONS_MURMUR64_*` directly.

**Comment by xmnlab** [comment]
@luizirber the points you described make sense, thanks for all de details! I am going to start on that now. thanks!

**Comment by xmnlab** [comment]
@luizirber should I also change `KmerMinHashBTree` signature?

File: sourmash/minhash.py
Comment (luizirber) [inline sourmash/minhash.py]: Nice trick, but can we have a regular `if/elif/else` here? =]

File: src/core/src/sketch/minhash.rs
Comment (luizirber) [inline src/core/src/sketch/minhash.rs]: how about using
```suggestion
            self.scaled,
```
here? `self.scaled` is still using `max_hash` in its impl, but we can fix that later (and avoid a mention to `self.max_hash` here =])

**Comment by luizirber** [comment]
> @luizirber should I also change `KmerMinHashBTree` signature?

Yes, please! `KmerMinHash` and `KmerMinHashBTree` have the same API but don't have a trait to connect them, so for now we need to change in both places... #1055 will fix it, but is not a priority.

File: src/core/src/sketch/minhash.rs
Comment (xmnlab) [reply src/core/src/sketch/minhash.rs]: @luizirber sounds good. could you add more details about that pls? :)

Comment (luizirber) [reply src/core/src/sketch/minhash.rs]: currently we have thousands of warnings in CI ([example](https://travis-ci.com/github/dib-lab/sourmash/jobs/376354282#L1605)) triggered because we are still using `.max_hash` internally. Eventually (for 4.0) we want to avoid using/documenting/suggesting `.max_hash` and focus on `.scaled` instead, including constructors (and hence why we deprecated `.max_hash` in 3.5 and generate all the warnings)

But that's all in the Python side, and this PR is for changing the Rust side. So, you don't need to worry about it if you don't want to, but I thought it was relevant to keep in mind =]

Comment (xmnlab) [reply src/core/src/sketch/minhash.rs]: I see .. thanks for the explanation :)

**Comment by xmnlab** [comment]
hey everyone, sorry for the delay here. I will try to finish this as soon as possible.

I have some issue with 2 tests, but not sure yet why ...

```
running 19 tests
test invalid_dna ... ok
test check_errors ... ok
test dayhoff ... ok
test hp ... ok
test load_save_minhash_sketches ... ok
test load_save_minhash_sketches_abund ... FAILED
test max_for_scaled ... ok
test merge ... ok
test merge_empty_scaled ... ok
test load_save_minhash_dna ... ok
test load_save_minhash_dayhoff ... ok
test load_save_minhash_hp ... ok
test similarity ... ok
test similarity_2 ... ok
test similarity_3 ... ok
test throws_error ... ok
test prop_merge ... FAILED
test oracle_mins_scaled ... ok
test oracle_mins ... ok

failures:

---- load_save_minhash_sketches_abund stdout ----
thread 'load_save_minhash_sketches_abund' panicked at 'called `Result::unwrap()` on an `Err` value: MismatchScaled', src/core/tests/minhash.rs:499:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- prop_merge stdout ----
proptest: FileFailurePersistence::SourceParallel set, but failed to find lib.rs or main.rs
thread 'prop_merge' panicked at 'called `Result::unwrap()` on an `Err` value: MismatchScaled', src/core/tests/minhash.rs:324:16
...
thread 'prop_merge' panicked at 'called `Result::unwrap()` on an `Err` value: MismatchScaled', src/core/tests/minhash.rs:324:16
thread 'prop_merge' panicked at 'Test failed: called `Result::unwrap()` on an `Err` value: MismatchScaled; minimal failing input: seq1 = "GGGTCC", seq2 = "CTTTGA"
        successes: 0
        local rejects: 0
        global rejects: 0
', src/core/tests/minhash.rs:299:1


failures:
    load_save_minhash_sketches_abund
    prop_merge

test result: FAILED. 17 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out

error: test failed, to rerun pass '--test minhash'
```

any tip about this?

**Comment by xmnlab** [comment]
it seems the problem is the new function `max_hash_for_scaled`

In:

```rust
pub fn max_hash_for_scaled(scaled: u64) -> u64 {
    match scaled {
        0 => 0,
        1 => u64::max_value(),
        _ => (u64::max_value() as f64 / scaled as f64) as u64,
    }
}

pub fn scaled_for_max_hash(max_hash: u64) -> u64 {
    match max_hash {
        0 => 0,
        _ => u64::max_value() / max_hash,
    }
}

let scaled = 10;
let max_hash = max_hash_for_scaled(scaled);

println!("max_hash (from scaled): {}", max_hash);
println!("scaled: {}", scaled);
println!("scaled (from max_hash): {}", scaled_for_max_hash(max_hash));
```

Output:

```
max_hash (from scaled): 1844674407370955264
scaled: 10
scaled (from max_hash): 9
```

@luizirber any recommendation about this?

**Comment by xmnlab** [comment]
I used the same approach for `scaled_for_max_hash` (converting to float) and it seems it worked, at least for the tests.

I tested locally the conversion between scaled_for_max_hash and max_hash_for_scaled and it seems it won't work in some cases (as the conversion to integer lost information). not sure if it is relevant for this  PR. 

let me know if I should take another direction :) thanks!

**Comment by xmnlab** [comment]
@ctb @luizirber finally it is ready for review. thanks for the patience!

**Comment by ctb** [comment]
hi @xmnlab thank you! I am not competent to review the rust code changes in detail, but everything looks good to me on the Python side (which you basically didn't change - yay :). @luizirber is working to a deadline so may take a few more days before he gets to review this, but let me be the first to say "good job!"

**Comment by xmnlab** [comment]
Thank you so much for the feedback @ctb <3
I am still learning Rust .. but I will be very happy to help if there is any other issue for beginner :)

**Comment by xmnlab** [comment]
hey everyone! just a friendly reminder about this PR :) let me know if I should address anything else here :) thanks!

File: sourmash/minhash.py
Comment (luizirber) [inline sourmash/minhash.py:248]: it makes sense that this works, but I've never seen it being used this way =]

**Review by luizirber — APPROVED** [review]
This looks great! Thanks @xmnlab!

**Comment by xmnlab** [comment]
thanks @ctb and @luizirber for all the patience and attention :)

**Comment by ctb** [comment]
thank YOU for all your work! I guess no PR is as simple as I initially
think :laugh:

**Pull request merged**
